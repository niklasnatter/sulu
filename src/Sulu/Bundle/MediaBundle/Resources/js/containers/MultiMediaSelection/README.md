The `MultiMediaSelection` component is based on the `MultiItemSelection` component and allows to assign media entries inside a form.
