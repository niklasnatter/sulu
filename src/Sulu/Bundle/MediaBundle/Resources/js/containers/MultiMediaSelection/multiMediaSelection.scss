.media-item {
    display: flex;
    height: 100%;
    align-items: center;

    .thumbnail-image {
        display: block;
        width: 25px;
        height: 25px;
    }

    .media-title {
        margin-left: 15px;
    }
}
