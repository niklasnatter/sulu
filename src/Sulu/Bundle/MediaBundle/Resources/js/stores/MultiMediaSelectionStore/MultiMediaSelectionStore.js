// @flow
import {action, computed, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import {arrayMove} from 'sulu-admin-bundle/components';
import {ResourceRequester} from 'sulu-admin-bundle/services';
import type {Media} from '../../types';

const MEDIA_RESOURCE_KEY = 'media';

export default class MultiMediaSelectionStore {
    @observable selectedMedia: Array<Media> = [];
    @observable loading: boolean = false;

    constructor(selectedMediaIds: ?Array<number>, locale: IObservableValue<string>) {
        if (selectedMediaIds && selectedMediaIds.length) {
            this.loadSelectedMedia(selectedMediaIds, locale);
        }
    }

    @action add(media: Media) {
        this.selectedMedia.push(media);
    }

    @action removeById(mediaId: number) {
        this.selectedMedia = this.selectedMedia.filter((media) => media.id !== mediaId);
    }

    @action move(oldItemIndex: number, newItemIndex: number) {
        this.selectedMedia = arrayMove(this.selectedMedia, oldItemIndex, newItemIndex);
    }

    @action setLoading(loading: boolean) {
        this.loading = loading;
    }

    @computed get selectedMediaIds(): Array<number> {
        return this.selectedMedia.map((media) => media.id);
    }

    @action loadSelectedMedia = (selectedMediaIds: Array<number>, locale: IObservableValue<string>) => {
        if (!selectedMediaIds || selectedMediaIds.length === 0) {
            this.selectedMedia = [];
            return;
        }

        this.setLoading(true);
        return ResourceRequester.getList(MEDIA_RESOURCE_KEY, {
            locale: locale.get(),
            ids: selectedMediaIds.join(','),
            limit: undefined, // TODO: Should be replaced by pagination
            page: 1,
        }).then(action((data) => {
            this.selectedMedia = data._embedded[MEDIA_RESOURCE_KEY];
            this.setLoading(false);
        }));
    };
}
