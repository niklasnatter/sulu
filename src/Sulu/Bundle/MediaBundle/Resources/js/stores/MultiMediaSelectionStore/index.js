// @flow
import MultiMediaSelectionStore from './MultiMediaSelectionStore';

export default MultiMediaSelectionStore;
