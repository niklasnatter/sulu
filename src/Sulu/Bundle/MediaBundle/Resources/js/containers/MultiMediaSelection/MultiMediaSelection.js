// @flow
import React, {Fragment} from 'react';
import {action, autorun, toJS, observable} from 'mobx';
import {observer} from 'mobx-react';
import equals from 'fast-deep-equal';
import {MultiItemSelection} from 'sulu-admin-bundle/components';
import {translate} from 'sulu-admin-bundle/utils';
import type {IObservableValue} from 'mobx';
import MultiMediaSelectionStore from '../../stores/MultiMediaSelectionStore';
import MultiMediaSelectionOverlay from '../MultiMediaSelectionOverlay';
import MimeTypeIndicator from '../../components/MimeTypeIndicator';
import multiMediaSelectionStyle from './multiMediaSelection.scss';
import type {Value} from './types';

type Props = {|
    disabled: boolean,
    locale: IObservableValue<string>,
    onChange: (selectedIds: Value) => void,
    value: Value,
|}

const THUMBNAIL_SIZE = 'sulu-25x25';

@observer
export default class MultiMediaSelection extends React.Component<Props> {
    static defaultProps = {
        disabled: false,
        value: {ids: []},
    };

    mediaSelectionStore: MultiMediaSelectionStore;
    changeDisposer: () => void;
    changeAutorunInitialized: boolean = false;

    @observable overlayOpen: boolean = false;

    constructor(props: Props) {
        super(props);

        const {locale, value} = this.props;

        this.mediaSelectionStore = new MultiMediaSelectionStore(value.ids, locale);
        this.changeDisposer = autorun(() => {
            const {onChange, value} = this.props;
            const loadedMediaIds = this.mediaSelectionStore.selectedMediaIds;

            if (!this.changeAutorunInitialized) {
                this.changeAutorunInitialized = true;
                return;
            }

            if (equals(toJS(value.ids), toJS(loadedMediaIds))) {
                return;
            }

            onChange({ids: loadedMediaIds});
        });
    }

    componentDidUpdate() {
        const {
            locale,
            value,
        } = this.props;

        const newSelectedIds = toJS(value.ids);
        const loadedSelectedIds = toJS(this.mediaSelectionStore.selectedMediaIds);

        newSelectedIds.sort();
        loadedSelectedIds.sort();
        if (!equals(newSelectedIds, loadedSelectedIds) && !this.mediaSelectionStore.loading) {
            this.mediaSelectionStore.loadSelectedMedia(newSelectedIds, locale);
        }
    }

    componentWillUnmount() {
        this.changeDisposer();
    }

    @action openMediaOverlay() {
        this.overlayOpen = true;
    }

    @action closeMediaOverlay() {
        this.overlayOpen = false;
    }

    getLabel(itemCount: number) {
        if (itemCount === 1) {
            return `1 ${translate('sulu_media.media_selected_singular')}`;
        } else if (itemCount > 1) {
            return `${itemCount} ${translate('sulu_media.media_selected_plural')}`;
        }

        return translate('sulu_media.select_media_plural');
    }

    handleRemove = (mediaId: number) => {
        this.mediaSelectionStore.removeById(mediaId);
    };

    handleSorted = (oldItemIndex: number, newItemIndex: number) => {
        this.mediaSelectionStore.move(oldItemIndex, newItemIndex);
    };

    handleOverlayOpen = () => {
        this.openMediaOverlay();
    };

    handleOverlayClose = () => {
        this.closeMediaOverlay();
    };

    handleOverlayConfirm = (selectedMedia: Array<Object>) => {
        selectedMedia.forEach((media) => this.mediaSelectionStore.add(media));
        this.closeMediaOverlay();
    };

    render() {
        const {locale, disabled} = this.props;

        const {
            loading,
            selectedMedia,
            selectedMediaIds,
        } = this.mediaSelectionStore;
        const label = (loading) ? '' : this.getLabel(selectedMedia.length);

        return (
            <Fragment>
                <MultiItemSelection
                    disabled={!!disabled}
                    label={label}
                    leftButton={{
                        icon: 'su-image',
                        onClick: this.handleOverlayOpen,
                    }}
                    loading={loading}
                    onItemRemove={this.handleRemove}
                    onItemsSorted={this.handleSorted}
                >
                    {selectedMedia.map((media, index) => {
                        return (
                            <MultiItemSelection.Item
                                id={media.id}
                                index={index + 1}
                                key={media.id}
                            >
                                <div className={multiMediaSelectionStyle.mediaItem}>
                                    {media.thumbnails[THUMBNAIL_SIZE]
                                        ? <img
                                            alt={media.title}
                                            className={multiMediaSelectionStyle.thumbnailImage}
                                            src={media.thumbnails[THUMBNAIL_SIZE]}
                                        />
                                        : <MimeTypeIndicator
                                            height={25}
                                            iconSize={16}
                                            mimeType={media.mimeType}
                                            width={25}
                                        />
                                    }
                                    <div className={multiMediaSelectionStyle.mediaTitle}>{media.title}</div>
                                </div>
                            </MultiItemSelection.Item>
                        );
                    })}
                </MultiItemSelection>
                <MultiMediaSelectionOverlay
                    excludedIds={selectedMediaIds}
                    locale={locale}
                    onClose={this.handleOverlayClose}
                    onConfirm={this.handleOverlayConfirm}
                    open={this.overlayOpen}
                />
            </Fragment>
        );
    }
}
