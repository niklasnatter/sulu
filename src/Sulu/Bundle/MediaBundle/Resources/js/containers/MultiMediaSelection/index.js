// @flow
import MultiMediaSelection from './MultiMediaSelection';
import type {Value} from './types';

export default MultiMediaSelection;
export type {Value};
