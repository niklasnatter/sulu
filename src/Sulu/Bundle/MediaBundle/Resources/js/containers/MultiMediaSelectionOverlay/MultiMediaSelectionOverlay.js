// @flow
import React from 'react';
import {action, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import {observer} from 'mobx-react';
import {DatagridStore} from 'sulu-admin-bundle/containers';
import MediaSelectionOverlay from '../MediaSelectionOverlay';

type Props = {|
    open: boolean,
    locale: IObservableValue<string>,
    excludedIds: Array<number>,
    onClose: () => void,
    onConfirm: (selectedMedia: Array<Object>) => void,
|};

@observer
export default class MultiMediaSelectionOverlay extends React.Component<Props> {
    static defaultProps = {
        excludedIds: [],
    };

    collectionId: IObservableValue<?string | number> = observable.box();
    excludedIdString: IObservableValue<string>;
    mediaDatagridStore: DatagridStore;
    collectionDatagridStore: DatagridStore;

    constructor(props: Props) {
        super(props);

        this.excludedIdString = observable.box(props.excludedIds.sort().join(','));

        this.mediaDatagridStore = MediaSelectionOverlay.createMediaDatagridStore(
            this.collectionId,
            this.excludedIdString,
            props.locale
        );
        this.collectionDatagridStore = MediaSelectionOverlay.createCollectionDatagridStore(
            this.collectionId,
            props.locale
        );
    }

    @action componentDidUpdate() {
        const newExcludedIdString = this.props.excludedIds.sort().join(',');

        if (this.excludedIdString.get() !== newExcludedIdString) {
            this.mediaDatagridStore.clear();
            this.excludedIdString.set(this.props.excludedIds.sort().join(','));
        }
    }

    componentWillUnmount() {
        if (this.mediaDatagridStore) {
            this.mediaDatagridStore.destroy();
        }

        if (this.collectionDatagridStore) {
            this.collectionDatagridStore.destroy();
        }
    }

    render() {
        const {
            onClose,
            onConfirm,
            open,
            locale,
        } = this.props;

        return (
            <MediaSelectionOverlay
                collectionDatagridStore={this.collectionDatagridStore}
                collectionId={this.collectionId}
                locale={locale}
                mediaDatagridStore={this.mediaDatagridStore}
                onClose={onClose}
                onConfirm={onConfirm}
                open={open}
            />
        );
    }
}
