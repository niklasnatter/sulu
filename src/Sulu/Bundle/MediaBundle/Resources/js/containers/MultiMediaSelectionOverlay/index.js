// @flow
import MultiMediaSelectionOverlay from './MultiMediaSelectionOverlay';

export default MultiMediaSelectionOverlay;
