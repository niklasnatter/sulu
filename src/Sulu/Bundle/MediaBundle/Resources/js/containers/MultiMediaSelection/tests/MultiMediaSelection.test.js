// @flow
import {mount, render, shallow} from 'enzyme';
import React from 'react';
import {extendObservable as mockExtendObservable, observable} from 'mobx';
import MultiMediaSelection from '../MultiMediaSelection';
import MultiMediaSelectionStore from '../../../stores/MultiMediaSelectionStore';
import MultiMediaSelectionOverlay from '../../MultiMediaSelectionOverlay';

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('../../MultiMediaSelectionOverlay', () => jest.fn(function() {
    return <div>single media selection overlay</div>;
}));

jest.mock('../../../stores/MultiMediaSelectionStore', () => jest.fn(function() {
    this.selectedMedia = [];
    this.selectedMediaIds = [];
    this.loadSelectedMedia = jest.fn();
}));

test('Render a MultiMediaSelection field', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function() {
        this.selectedMedia = [
            {
                id: 1,
                title: 'Media 1',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
            {
                id: 2,
                title: 'Media 2',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
            {
                id: 3,
                title: 'Media 3',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
        ];
        this.selectedMediaIds = [1, 2, 3];
    });

    expect(render(
        <MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />
    )).toMatchSnapshot();
});

test('The MultiMediaSelection should have 3 child-items', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function() {
        this.selectedMedia = [
            {
                id: 1,
                title: 'Media 1',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
            {
                id: 2,
                title: 'Media 2',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
            {
                id: 3,
                title: 'Media 3',
                thumbnails: {
                    'sulu-25x25': 'http://lorempixel.com/25/25',
                },
            },
        ];
        this.selectedMediaIds = [1, 2, 3];
    });

    const mediaSelection = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />
    );

    expect(mediaSelection.find('Item').length).toBe(3);
});

test('Clicking on the "add media" button should open up an overlay', () => {
    const mediaSelection = mount(<MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />);

    expect(mediaSelection.find(MultiMediaSelectionOverlay).prop('open')).toEqual(false);
    mediaSelection.find('.button.left').simulate('click');
    expect(mediaSelection.find(MultiMediaSelectionOverlay).prop('open')).toEqual(true);
});

test('Should remove media from the selection store', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function() {
        this.selectedMedia = [];
        this.selectedMediaIds = [];
        this.removeById = jest.fn();
    });

    const mediaSelectionInstance = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />
    ).instance();

    mediaSelectionInstance.handleRemove(1);
    expect(mediaSelectionInstance.mediaSelectionStore.removeById).toBeCalledWith(1);
});

test('Should move media inside the selection store', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function() {
        this.selectedMedia = [];
        this.selectedMediaIds = [];
        this.move = jest.fn();
    });

    const mediaSelectionInstance = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />
    ).instance();

    mediaSelectionInstance.handleSorted(1, 3);
    expect(mediaSelectionInstance.mediaSelectionStore.move).toBeCalledWith(1, 3);
});

test('Should add the selected medias to the selection store on confirm', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function() {
        this.selectedMedia = [];
        this.selectedMediaIds = [];
        this.add = jest.fn();
    });

    const thumbnails = {
        'sulu-240x': 'http://lorempixel.com/240/100',
        'sulu-25x25': 'http://lorempixel.com/25/25',
    };
    const mediaSelectionInstance = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={jest.fn()} />
    ).instance();

    mediaSelectionInstance.openMediaOverlay();
    mediaSelectionInstance.handleOverlayConfirm([
        {
            id: 1,
            title: 'Title 1',
            mimeType: 'image/png',
            size: 12345,
            url: 'http://lorempixel.com/500/500',
            thumbnails: thumbnails,
        },
        {
            id: 2,
            title: 'Title 2',
            mimeType: 'image/jpeg',
            size: 54321,
            url: 'http://lorempixel.com/500/500',
            thumbnails: thumbnails,
        },
    ]);
    expect(mediaSelectionInstance.mediaSelectionStore.add.mock.calls[0][0].id).toBe(1);
    expect(mediaSelectionInstance.mediaSelectionStore.add.mock.calls[0][0].title).toBe('Title 1');
    expect(mediaSelectionInstance.mediaSelectionStore.add.mock.calls[1][0].id).toBe(2);
    expect(mediaSelectionInstance.mediaSelectionStore.add.mock.calls[1][0].title).toBe('Title 2');
    expect(mediaSelectionInstance.overlayOpen).toBe(false);
});

test('Should call the onChange handler if selection store changes', () => {
    // $FlowFixMe
    MultiMediaSelectionStore.mockImplementationOnce(function(selectedIds) {
        mockExtendObservable(this, {
            selectedMedia: selectedIds.map((id) => {
                return {id, thumbnails: {}};
            }),
            get selectedMediaIds() {
                return this.selectedMedia.map((media) => media.id);
            },
        });
    });

    const changeSpy = jest.fn();

    const mediaSelectionInstance = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={changeSpy} value={{ids: [55]}} />
    ).instance();

    mediaSelectionInstance.mediaSelectionStore.selectedMedia.push({id: 99, thumbnails: {}});
    expect(changeSpy).toBeCalledWith({ids: [55, 99]});

    mediaSelectionInstance.mediaSelectionStore.selectedMedia.splice(0, 1);
    expect(changeSpy).toBeCalledWith({ids: [99]});
});

test('Should not call the onChange callback if the component props change', () => {
    const changeSpy = jest.fn();

    const mediaSelection = shallow(
        <MultiMediaSelection locale={observable.box('en')} onChange={changeSpy} value={{ids: [55]}} />
    );

    mediaSelection.setProps({disabled: true});
    expect(changeSpy).not.toBeCalled();
});

test('Pass correct props to MultiItemSelection component', () => {
    const mediaSelection = mount(
        <MultiMediaSelection disabled={true} locale={observable.box('en')} onChange={jest.fn()} />
    );

    expect(mediaSelection.find('MultiItemSelection').prop('disabled')).toEqual(true);
});
