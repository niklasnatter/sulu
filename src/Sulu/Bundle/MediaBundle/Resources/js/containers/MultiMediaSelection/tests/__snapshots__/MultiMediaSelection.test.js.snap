// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a MultiMediaSelection field 1`] = `
Array [
  <div
    class="multiItemSelectionClass"
  >
    <div
      class="header"
    >
      <button
        class="button left"
        type="button"
      >
        <span
          aria-label="su-image"
          class="su-image"
        />
      </button>
      <div
        class="label"
      >
        3 sulu_media.media_selected_plural
      </div>
    </div>
    <ul
      class="list"
    >
      <li
        class="listElement"
      >
        <div
          class="item"
        >
          <span
            class="dragHandle sortable"
          >
            <span
              aria-label="su-more"
              class="su-more"
            />
            <span
              class="index"
            >
              1
            </span>
          </span>
          <div
            class="content"
          >
            <div
              class="mediaItem"
            >
              <img
                alt="Media 1"
                class="thumbnailImage"
                src="http://lorempixel.com/25/25"
              />
              <div
                class="mediaTitle"
              >
                Media 1
              </div>
            </div>
          </div>
          <button
            class="removeButton"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </li>
      <li
        class="listElement"
      >
        <div
          class="item"
        >
          <span
            class="dragHandle sortable"
          >
            <span
              aria-label="su-more"
              class="su-more"
            />
            <span
              class="index"
            >
              2
            </span>
          </span>
          <div
            class="content"
          >
            <div
              class="mediaItem"
            >
              <img
                alt="Media 2"
                class="thumbnailImage"
                src="http://lorempixel.com/25/25"
              />
              <div
                class="mediaTitle"
              >
                Media 2
              </div>
            </div>
          </div>
          <button
            class="removeButton"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </li>
      <li
        class="listElement"
      >
        <div
          class="item"
        >
          <span
            class="dragHandle sortable"
          >
            <span
              aria-label="su-more"
              class="su-more"
            />
            <span
              class="index"
            >
              3
            </span>
          </span>
          <div
            class="content"
          >
            <div
              class="mediaItem"
            >
              <img
                alt="Media 3"
                class="thumbnailImage"
                src="http://lorempixel.com/25/25"
              />
              <div
                class="mediaTitle"
              >
                Media 3
              </div>
            </div>
          </div>
          <button
            class="removeButton"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </li>
    </ul>
  </div>,
  <div>
    single media selection overlay
  </div>,
]
`;
